src/runtime/virtcontainers/network_linux.go

```diff
@@ -516,7 +516,13 @@ func xConnectVMNetwork(ctx context.Context, endpoint Endpoint, h Hypervisor) err
 	queues := 0
 	caps := h.Capabilities(ctx)
 	if caps.IsMultiQueueSupported() {
-		queues = int(h.HypervisorConfig().NumVCPUs)
+		if h.HypervisorConfig().QueueNums == 0 {
+            		queues = int(h.HypervisorConfig().NumVCPUs)
+       		} else if int(h.HypervisorConfig().NumVCPUs) > int(h.HypervisorConfig().QueueNums) {
+            		queues = int(h.HypervisorConfig().QueueNums)
+       		} else{
+            		queues = int(h.HypervisorConfig().NumVCPUs)
+       		}
```

Comment:
The logic looks a bit fuzzy to me, especially when `QueueNums == 0`... how is this happening ? Because the user passed the annotations with a value of 0 ? Because the user didn't pass the annotation and it supposedly defaults to 0 ?

I think that `QueueNums` should belong to [1..NumVCPUs]. Any value outside this range should cause an error or at least a visible warning. The default value should be NumVCPUs to match the existing behavior.

It seems all of this would be better located in the [`Valid()` function of the `hypervisor` module](https://github.com/kata-containers/kata-containers/blob/4443bb68a45f3d1eb169b6c1cb223bfbfe9d09cb/src/runtime/virtcontainers/hypervisor.go#L529) to keep `xConnectVMNetwork()` simple.

Also, annotations are API that must be documented in `src/runtime/virtcontainers/documentation/api/1.0/api.md`.

Reply:
You are right, I use `QueueNums == 0`  that means the user didn't pass the annotation, but as you can see this may be not suitable. I will use a default value and achieve it in `func (conf *HypervisorConfig) Valid() error`.

Reply:
Actually you should only have `queues = int(h.HypervisorConfig().NumQueues)` in this function. All the checking should go to `Valid()`. Also, even if it is generally a bad idea to have more queues than processing power, we shouldn't silently change a user requested setting. Personally I would just print a warning.

Reply:
I follow your suggestions that have move the checking go to `Valid()`.
But I think it better to change the number of queues when more queues than processing power, if I set `NumQueues = 3` and `NumVCPUs = 2`, it can be seen as follows:
`Channel parameters for tap0:
Pre-set maximums:
RX:             0
TX:             0
Other:          0
Combined:       3
Current hardware settings:
RX:             0
TX:             0
Other:          0
Combined:       2
`
Actually, the max number of queues should be 2 not 3.

Reply:
Fine with me. So now, just fold everything into a single patch. Ensure you don't have spaces in the indention. Fix the From: and Signed-off-by:, force-push to GH and we should be good to go @Gezx . :smile:

Reply:
@Gezx I tend to agree with @gkurz here. Notably, if there are multiple containers with multiple annotations, and the second container adds CPUs, then you may end up in a reasonable configuration.

You could also recognize a special value, e.g. `-1`, to mean "use number of VCPUs".